invenio_rdm_records/secret_links/models.py

```diff
@@ -0,0 +1,172 @@
+
+from invenio_db import db
+from itsdangerous import BadData
+from sqlalchemy_utils import UUIDType
+
+from .permissions import LinkNeed
+from .serializers import SecretLinkSerializer, TimedSecretLinkSerializer
+from .signals import link_created, link_revoked
+
+SUPPORTED_DIGEST_ALGORITHMS = ("HS256", "HS512")
+
+
+class SecretLink(db.Model):
+    """Secret links for sharing permissions on records."""
+
+    __tablename__ = "rdm_records_secret_links"
+
+    id = db.Column(UUIDType, primary_key=True, default=uuid.uuid4)
+    """Secret link ID."""
+
+    token = db.Column(db.Text, nullable=False)
+    """Secret token for link.
+
+    In contrast to Zenodo-AccessRequests, we don't store encrypted tokens, to
+    enable querying the SecretLink table by tokens and improve performance.
+    In case that the database is leaked, all secret links should be revoked.
+    """
+
+    created = db.Column(
+        db.DateTime, nullable=False, default=datetime.utcnow, index=True
+    )
+    """Creation timestamp."""
+
+    expires_at = db.Column(db.DateTime, nullable=True)
+    """Expiration date."""
+
+    permission_level = db.Column(db.String, nullable=False, default="")
+    """Permission level of the link."""
+
+    def allows(self, action):
+        """Check if the secret link covers the specified permission."""
+        # TODO permission hierarchy
+        return self.permission_level == action
+
+    def revoke(self, commit=False):
+        """Revoke (i.e. delete) this secret link."""
+        db.session.delete(self)
+        if commit:
+            db.session.commit()
+
+        link_revoked.send(self)
+
+    @property
+    def need(self):
+        """Get the needs provided by the link."""
+        return LinkNeed(str(self.id))
+
+    @property
+    def extra_data(self):
+        """Load token data stored in token (ignores expiry date of tokens)."""
+        if self.token:
+            return self.load_token(self.token, force=True)["data"]
+
+        return None
+
+    @property
+    def is_expired(self):
+        """Determine if link is expired."""
+        if self.expires_at:
+            return datetime.utcnow() > self.expires_at
+
+        return False
+
+    @classmethod
+    def get_by_token(cls, token):
+        """Get the SecretLink referenced by the specified token."""
+        data = cls.load_token(token)
+        if data:
+            link = cls.query.get(data["id"])
+            return link
+        else:
+            return None
+
+    @classmethod
+    def create(
+        cls,
+        permission_level,
+        extra_data=dict(),
+        expires_at=None,
+        commit=False,
+    ):
+        """Create a new secret link."""
+        is_date = isinstance(expires_at, date)
+        is_datetime = isinstance(expires_at, datetime)
+        if is_date and not is_datetime:
+            expires_at = datetime.combine(expires_at, datetime.min.time())
+
+        with db.session.begin_nested():
+            obj = cls(
```

Comment:
Here I would simply do:
```python
id_ = uuid.uuid4()
obj = cls(
    id=id_
    ....,
    token=serializer.create_token(id_, extra_data)
)
```

There's a super low chance of collission and we also do it other places (i.e. generating the uuid in the app)

Reply:
done; removed the now no longer required nested sessions